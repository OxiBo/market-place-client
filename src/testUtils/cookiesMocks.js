import { Cookies } from "react-cookie";
export const setMyCookies = (token, type, name, id) => {
  const cookiesMocks = new Cookies();
  cookiesMocks.set("token", token);
  cookiesMocks.set("type", type);
  cookiesMocks.set("name", name);
  cookiesMocks.set("id", id);
  return cookiesMocks;
};
