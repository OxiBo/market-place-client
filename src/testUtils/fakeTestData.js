import casual from "casual";

const fakeBuyer = () => ({
  __typename: "User",
  id: "4234",
  name: casual.name,
  email: casual.email,
  age: 21,
  image:
    "https://res.cloudinary.com/di0hg10hd/image/upload/v1597378083/sickfits/qekc7yfql14ligautxzq.jpg",
  orderItems: [],
  reviews: [],
  type: "BUYER",
});

const fakeOrderItems = () => [
  {
    __typename: "OrderItem",
    id: "123asd",
    image: "",
    count: 2,
    name: "fake order item 0",
    price: 10,
    product: {
      __typename: "Product",
      id: "456w",
      name: "fake order item 0",
      seller: {
        id: "zxcv",
        name: "EbayTest",
        __typename: "Seller",
      },
    },
    reviewed: null,
    order: {
      id: "asdfg",
      finishedAt: "2020-09-02T02:34:55.835Z",
      __typename: "Order",
    },
  },

  {
    __typename: "OrderItem",
    id: "123vbn",
    image: "",
    count: 3,
    name: "fake order item 1",
    price: 10,
    product: {
      __typename: "Product",
      id: "456",
      name: "fake order item 1",
      seller: {
        id: "zxcv123",
        name: "AmazonTest",
        __typename: "Seller",
      },
    },
    reviewed: null,
    order: {
      id: "asdfgsss",
      finishedAt: "2020-09-02T02:34:55.835Z",
      __typename: "Order",
    },
  },
];

export { fakeBuyer, fakeOrderItems };
