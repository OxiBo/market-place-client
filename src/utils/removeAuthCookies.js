const removeAuthCookies = (cookies) => {
  cookies.remove("token", { path: "/" });
  cookies.remove("name", { path: "/" });
  cookies.remove("id", { path: "/" });
  cookies.remove("type", { path: "/" });
};

export default removeAuthCookies;
