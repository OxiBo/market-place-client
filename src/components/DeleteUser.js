import React from "react";
import { useHistory } from "react-router-dom";
import { Mutation } from "react-apollo";
import Button from "./styled/Button";
import { DELETE_USER, DELETE_SELLER } from "../utils/serverOperations";
import removeAuthCookies from "../utils/removeAuthCookies";

const DeleteUser = ({ type, cookies }) => {
  //   console.log(cookies);
  const history = useHistory();
  return (
    <Mutation
      mutation={type === "BUYER" ? DELETE_USER : DELETE_SELLER}
      onCompleted={() => {
        removeAuthCookies(cookies);
        history.push("/");
      }}
    >
      {(deleteUser) => (
        <Button
          onClick={() => {
              // TODO make the confirm window - react modal
            if (window.confirm("Are you sure you want to delete your profile?")) {
              deleteUser(cookies).catch((err) => {
                console.error(err);
                alert(err.message);
              });
            }
          }}
        >
          {" "}
          Delete Profile
        </Button>
      )}
    </Mutation>
  );
};

export default DeleteUser;
