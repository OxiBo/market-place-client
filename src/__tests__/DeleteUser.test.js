import React from "react";
import { useHistory } from "react-router-dom";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import { act } from "react-dom/test-utils";
import { ApolloConsumer } from "react-apollo";
import { MemoryRouter, withRouter } from "react-router-dom"; // https://html.developreference.com/article/13543175/React+Router+Invariant+Violation+test+fails+with+%3CLink%3E+outside+%3CRouter%3E+%5Bduplicate%5D
// why MemoryRouter - https://github.com/enzymejs/enzyme/issues/1112#issuecomment-447643271
import DeleteUser from "../components/DeleteUser";
// import { CURRENT_USER_QUERY } from "../components/User";
// import { MockedProvider } from "react-apollo/test-utils";
import { MockedProvider } from "@apollo/react-testing";
// mocking withCookies - https://github.com/reactivestack/cookies/blob/master/packages/react-cookie/src/__tests__/withCookies-test.js
import { CookiesProvider, withCookies, Cookies } from "react-cookie";
import { setMyCookies } from "../testUtils/cookiesMocks";
import {
  FETCH_USER_PROFILE,
  DELETE_SELLER,
  DELETE_USER,
} from "../utils/serverOperations";
import { fakeBuyer, fakeDeletedBuyer } from "../testUtils/fakeTestData";
import removeAuthCookies from "../utils/removeAuthCookies";

const signedInBuyerProfile = [
  {
    request: { query: FETCH_USER_PROFILE },
    result: {
      data: {
        meUser: fakeBuyer(),
      },
    },
  },
];

const deletedUser = [
  {
    request: { query: DELETE_USER },
    result: {
      data: {
        deleteUser: fakeDeletedBuyer(),
      },
    },
  },
  //   {
  //     request: { mutation: DELETE_SELLER },
  //     result: {
  //       data: {
  //         deleteUser: "abc123",
  //       },
  //     },
  //   },
];

// const historyMock = jest.mock("react-router-dom", () => ({
//   useHistory: () => ({
//     push: jest.fn(),
//   }),
// }));

const historyMock = { push: jest.fn(), location: {}, listen: jest.fn() };
// const historyMock  = jest.mock('react-router-dom', () => ({
//     ...jest.requireActual('react-router-dom'),
//     useHistory: () => ({
//       push: jest.fn()
//     })
//   }));

let myCookies, wrapper, deleteUser, apolloClient;

beforeEach(async () => {
  myCookies = setMyCookies(
    "token_lskdjflskdjflskdjflskdf",
    "BUYER",
    "Test",
    "lskdjflskdjflskdjflskdf"
  );
  const cookies = new Cookies();
  wrapper = mount(
    <CookiesProvider cookies={myCookies}>
      <MemoryRouter>
        <MockedProvider mocks={[...signedInBuyerProfile, ...deletedUser]}>
          <ApolloConsumer>
            {(client) => {
              apolloClient = client;
              return (
                <DeleteUser
                  type={"BUYER"}
                  history={historyMock}
                  cookies={cookies}
                />
              );
            }}
          </ApolloConsumer>
        </MockedProvider>
      </MemoryRouter>
    </CookiesProvider>
  );
  await act(async () => wait());
  wrapper.update();
  deleteUser = wrapper.find("DeleteUser");
  //   pureComponent = buyerProfile.instance();
});

describe("<DeleteUser />", () => {
  it("renders the delete button and matches the snapshot", () => {
    // const cookies = new Cookies();
    // const history = useHistory();
    // mocking withCookies - https://github.com/reactivestack/cookies/blob/master/packages/react-cookie/src/__tests__/withCookies-test.js
    // const Component = withCookies(NavBar);

    // console.log(deleteUser.debug());
    expect(toJSON(deleteUser)).toMatchSnapshot();
  });

//   it("deletes user and removes cookies on button click", async () => {
//     const deleteButton = deleteUser.find("button");
//     console.log(deleteButton.debug());

//     deleteButton.simulate("click");
//     await act(async () => wait());
//     wrapper.update();
//     const res = await apolloClient.query({query: FETCH_USER_PROFILE})
//     console.log(res)
//     console.log(wrapper.debug());

//     // expect(historyMock.push.mock).toHaveBeenCalled();
//     console.log(historyMock.push.mock);
//     expect(historyMock.push.mock.calls[0][0]).toEqual("/");
//     //    expect(deleteUser).toHaveBeenCalled()
//     // expect(loginButton.text()).toContain("Login");
//   });
});
