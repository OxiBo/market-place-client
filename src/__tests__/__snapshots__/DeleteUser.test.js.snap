// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeleteUser /> renders the delete button and matches the snapshot 1`] = `
<DeleteUser
  cookies={
    Cookies {
      "HAS_DOCUMENT_COOKIE": true,
      "changeListeners": Array [],
      "cookies": Object {},
    }
  }
  history={
    Object {
      "listen": [MockFunction],
      "location": Object {},
      "push": [MockFunction],
    }
  }
  type="BUYER"
>
  <Mutation
    mutation={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": Object {
              "kind": "Name",
              "value": "deleteUser",
            },
            "operation": "mutation",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "deleteUser",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 61,
          "start": 0,
        },
      }
    }
    onCompleted={[Function]}
  >
    <styled.button
      onClick={[Function]}
    >
      <button
        className="sc-AxjAm coAxVD"
        onClick={[Function]}
      >
         
        Delete Profile
      </button>
    </styled.button>
  </Mutation>
</DeleteUser>
`;
